src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,274 @@
+	opts WideQueryOptions,
+) BaseResults {
+	batchSize := opts.BatchSize
+	results := &wideResults{
+		nsID:        namespaceID,
+		idPool:      idPool,
+		batchSize:   batchSize,
+		idsOverflow: make([]ident.ID, 0, batchSize),
+		batch: &ident.IDBatch{
+			IDs: make([]ident.ID, 0, batchSize),
+		},
+		batchCh: collector,
+		shards:  opts.ShardsQueried,
+	}
+
+	if len(opts.ShardsQueried) > 0 {
+		// Only apply filter if there are shards to filter against.
+		results.shardFilter = shardFilter
+	}
+
+	return results
+}
+
+func (r *wideResults) EnforceLimits() bool { return false }
+
+func (r *wideResults) AddDocuments(batch []doc.Document) (int, int, error) {
+	var size, totalDocsCount int
+	r.RLock()
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.RUnlock()
+
+	if r.closed || r.pastLastShard {
+		return size, totalDocsCount, ErrWideQueryResultsExhausted
+	}
+
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.Unlock()
+
+	if err != nil && err != ErrWideQueryResultsExhausted {
+		// NB: if exhausted, drain the current batch and overflows.
+		return size, totalDocsCount, err
+	}
+
+	release := len(r.batch.IDs) == r.batchSize
+	if release {
+		r.releaseAndWait()
+		r.releaseOverflow()
+	}
+
+	r.RLock()
+	size = r.size
+	r.RUnlock()
+
+	return size, totalDocsCount, err
+}
+
+func (r *wideResults) addDocumentsBatchWithLock(batch []doc.Document) error {
+	for i := range batch {
+		if err := r.addDocumentWithLock(batch[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *wideResults) addDocumentWithLock(d doc.Document) error {
+	if len(d.ID) == 0 {
+		return errUnableToAddResultMissingID
+	}
+
+	var tsID ident.ID = ident.BytesID(d.ID)
+
+	// Need to apply filter if set first.
+	if r.shardFilter != nil {
+		filteringShard := r.shards[r.shardIdx]
+		documentShard, documentShardOwned := r.shardFilter(tsID)
+		// NB: Check to see if shard is exceeded first (to short circuit earlier if
+		// the current shard is not owned by this node, but shard exceeds filtered).
+		if filteringShard > documentShard {
+			// this document is from a shard lower than the next shard allowed.
+			return nil
+		}
+
+		for filteringShard < documentShard {
+			// this document is from a shard higher than the next shard allowed;
+			// advance to the next shard, then try again.
+			r.shardIdx++
+			if r.shardIdx >= len(r.shards) {
+				// shard is past the final shard allowed by filter, no more results
+				// will be accepted.
+				r.pastLastShard = true
+				return ErrWideQueryResultsExhausted
+			}
+
+			filteringShard = r.shards[r.shardIdx]
+			if filteringShard > documentShard {
+				// this document is from a shard lower than the next shard allowed.
+				return nil
+			}
+		}
+
+		if !documentShardOwned {
+			return nil
+		}
+	}
+
+	r.size++
+	r.totalDocsCount++
+
+	// Pool IDs after filter is passed.
+	tsID = r.idPool.Clone(tsID)
+	if len(r.batch.IDs) < r.batchSize {
```

Comment:
Once we hit the batch size, how do we restart at the very ID we missed? I see all the shard filtering but I don't understand how the very next batch kicks off properly.

Reply:
They get added to `r.idsOverflow`, which is bounded by `len(batch)` from [here](https://github.com/m3db/m3/pull/2632/files/39502e996424f6c114ebc94c34341387bddb6b78#diff-23fd2a66d7fff72638b2f6e385d479f3c86aae924d3503368894e917fb5a0b92R131); this gets drained if possible in `releaseOverflow`, until if it's not large enough for a fresh batch, then it seeds the next `r.batch` with the overflow